plugins/bkpr/recorder.c

```diff
@@ -0,0 +1,545 @@
+#include <plugins/bkpr/onchain_fee.h>
+#include <plugins/bkpr/recorder.h>
+
+static struct chain_event *stmt2chain_event(const tal_t *ctx, struct db_stmt *stmt)
+{
+	struct chain_event *e = tal(ctx, struct chain_event);
+	e->db_id = db_col_u64(stmt, "id");
+	e->acct_db_id = db_col_u64(stmt, "account_id");
+
+	e->tag = db_col_strdup(e, stmt, "tag");
+
+	db_col_amount_msat(stmt, "credit", &e->credit);
+	db_col_amount_msat(stmt, "debit", &e->debit);
+	db_col_amount_msat(stmt, "output_value", &e->output_value);
+
+	e->currency = db_col_strdup(e, stmt, "currency");
+	e->timestamp = db_col_u64(stmt, "timestamp");
+	e->blockheight = db_col_int(stmt, "blockheight");
+
+	db_col_txid(stmt, "utxo_txid", &e->outpoint.txid);
+	e->outpoint.n = db_col_int(stmt, "outnum");
+
+	if (!db_col_is_null(stmt, "payment_id")) {
+		e->payment_id = tal(e, struct sha256);
+		db_col_sha256(stmt, "payment_id", e->payment_id);
+	} else
+		e->payment_id = NULL;
+
+	if (!db_col_is_null(stmt, "spending_txid")) {
+		e->spending_txid = tal(e, struct bitcoin_txid);
+		db_col_txid(stmt, "spending_txid", e->spending_txid);
+	} else
+		e->spending_txid = NULL;
+
+	return e;
+}
+
+static struct channel_event *stmt2channel_event(const tal_t *ctx, struct db_stmt *stmt)
+{
+	struct channel_event *e = tal(ctx, struct channel_event);
+
+	e->db_id = db_col_u64(stmt, "id");
+	e->acct_db_id = db_col_u64(stmt, "account_id");
+
+	e->tag = db_col_strdup(e, stmt, "tag");
+
+	db_col_amount_msat(stmt, "credit", &e->credit);
+	db_col_amount_msat(stmt, "debit", &e->debit);
+	db_col_amount_msat(stmt, "fees", &e->fees);
+
+	e->currency = db_col_strdup(e, stmt, "currency");
+	db_col_sha256(stmt, "payment_id", &e->payment_id);
+	e->part_id = db_col_int(stmt, "part_id");
+	e->timestamp = db_col_u64(stmt, "timestamp");
+
+	return e;
+}
+
+struct chain_event **account_get_chain_events(const tal_t *ctx,
+					      struct db *db,
+					      struct account *acct)
+{
+	struct db_stmt *stmt;
+	struct chain_event **results;
+
+	stmt = db_prepare_v2(db, SQL("SELECT"
+				     "  id"
+				     ", account_id"
+				     ", tag"
+				     ", credit"
+				     ", debit"
+				     ", output_value"
+				     ", currency"
+				     ", timestamp"
+				     ", blockheight"
+				     ", utxo_txid"
+				     ", outnum"
+				     ", spending_txid"
+				     ", payment_id"
+				     " FROM chain_events"
+				     " WHERE account_id = ?;"));
+
+	db_bind_int(stmt, 0, acct->db_id);
+	db_query_prepared(stmt);
+
+	results = tal_arr(ctx, struct chain_event *, 0);
+	while (db_step(stmt)) {
+		struct chain_event *e = stmt2chain_event(results, stmt);
+		tal_arr_expand(&results, e);
+	}
+	tal_free(stmt);
+
+	return results;
+}
+
+static struct chain_event *find_chain_event(const tal_t *ctx,
+					    struct db *db,
+					    const struct account *acct,
+					    const struct bitcoin_outpoint *outpoint,
+					    const struct bitcoin_txid *spending_txid)
+
+{
+	struct db_stmt *stmt;
+	struct chain_event *e;
+
+	if (spending_txid) {
```

Comment:
I see, that might require a bit more normalization of the conditions, and become longer, but it should work. It's a tradeoff though, and if it gets too long and wonky it might definitely not be worth doing :+1: